practice/rain.py

```diff
@@ -0,0 +1,193 @@
+"""Rain database processor.
+
+Accesses City of Portland rain total database and returns various information
+about rain totals.
+"""
+import urllib.request
+import re
+import os.path
+import statistics
+
+
+def get_daily_totals(filename):
+    """Return a dict of days and daily amounts, as rows from the file."""
+    totals = {}
+    with open(filename) as f:
+        for i in range(11):
+            _ = f.readline()
+        for line in f:
+            day_stats = line.split()[0:2]
+            if day_stats[1] != '-':  # skip entries if there is missing data
+                totals.update({day_stats[0]: int(day_stats[1])})
+    return totals
```

Comment:
I suggest breaking apart your file reading from any file processing. Then you can very easily test if your processing (line skipping, splitting, and missing data processing) works.

One function just reads the file into a list of strings with `f.readlines()`. Another function does all of the other steps using an explicit list of lines.

You can't test the file reading in doctests, but you can test the processing, which is the part that could have more bugs.